signer/src/api/new_block.rs

```diff
@@ -112,34 +143,266 @@ pub async fn new_block_handler(state: State<ApiState<impl Context>>, body: Strin
         // in a second.
-        if let Err(error) = res {
+        if let Err(Error::SqlxQuery(error)) = res {
             tracing::error!(%error, "Got an error when writing event to database");
             return StatusCode::INTERNAL_SERVER_ERROR;
+        // If we got an error processing the event, we log the error and
+        // return a success status code so that the node does not retry the
+        // webhook. We rely on the redundancy of the other sBTC signers to
+        // ensure that the update is sent to Emily.
+        } else if let Err(error) = res {
+            tracing::error!(%error, "Got an error when processing event");
         }
     }
 
-    // Propagate the new block to Emily.
+    // Send the updates to Emily.
+    let emily_client = api.ctx.get_emily_client();
     let chainstate = Chainstate::new(block_id.to_string(), new_block_event.block_height);
-    if let Err(result) = api.ctx.get_emily_client().set_chainstate(chainstate).await {
-        tracing::warn!(%result, "Failed to set chainstate in emily");
+    let futures = vec![
+        emily_client
+            .update_deposits(completed_deposits)
+            .map(UpdateResult::Deposit)
+            .boxed(),
+        emily_client
+            .update_withdrawals(updated_withdrawals)
+            .map(UpdateResult::Withdrawal)
+            .boxed(),
+        emily_client
+            .create_withdrawals(created_withdrawals)
+            .map(UpdateResult::CreatedWithdrawal)
+            .boxed(),
+        emily_client
+            .set_chainstate(chainstate)
+            .map(UpdateResult::Chainstate)
+            .boxed(),
+    ];
+    // TODO: Ideally, we would use `futures::future::join_all` here, but Emily
+    // randomly returns a `VersionConflict` error when we send multiple
+    // requests that may update the chainstate.
+    // let results = futures::future::join_all(futures).await;
+
+    // Log any errors that occurred while updating Emily.
+    // We don't return a non-success status code here because we rely on
+    // the redundancy of the other sBTC signers to ensure that the update
+    // is sent to Emily.
+    for future in futures {
+        let result = future.await;
+        if let UpdateResult::Chainstate(Err(error)) = result {
+            tracing::warn!(%error, "Failed to set chainstate in emily");
+        } else if let UpdateResult::Deposit(Err(error)) = result {
+            tracing::warn!(%error, "Failed to update deposits in emily");
+        } else if let UpdateResult::Withdrawal(Err(error)) = result {
+            tracing::warn!(%error, "Failed to update withdrawals in emily");
+        } else if let UpdateResult::CreatedWithdrawal(results) = result {
+            for result in results {
+                if let Err(error) = result {
+                    tracing::warn!(%error, "Failed to create withdrawals in emily");
+                }
+            }
+        }
```

Comment:
nit: why not a match statement here?

Reply:
ah, you caught me! I went for the greedy route to avoid adding the extra branch for successful cases :D but I’ll happily switch to the match statement. Thanks for the nudge!